### ISSUE TITLE
Would it be possible to support the #[deprecated] attribute?

### ISSUE THREAD

**Issue by Takashiidobe** [extendr/extendr#382]

**Comment by yutannihilation** [comment]
Thanks!

**Comment by yutannihilation** [comment]
Just a quick note. The short answer to the title question should be "yes". 

I get this `TokenStream` for `#[deprecated]`.

`TokenStream [Punct { ch: '#', spacing: Alone, span: #0 bytes(1363..1364) }, Group { delimiter: Bracket, stream: TokenStream [Ident { ident: "deprecated", span: #0 bytes(1365..1375) }], span: #0 bytes(1364..1376) }, ...]`

Maybe we can add some additional processing here:

https://github.com/extendr/extendr/blob/98f5fe2837af5c43e5542b547533422706728c95/extendr-macros/src/lib.rs#L75

**Comment by yutannihilation** [comment]
Yeah, that's the question. I think this kind of feature makes sense if it's a common usage to provide both Rust interface and R package from one Rust code (disclaimer: I prefer writing some wrapper code on R's side anyway, so I'm not the right person here).

**Comment by clauswilke** [comment]
Regardless of whether this gets added or not, I don't think you should ever manually change the auto-generated R bindings.

One way you could solve the problem without manual intervention every time is to rename the function that gets exported by Rust, e.g. to `multiply_deprecated()`, and then write an R wrapper that creates the deprecation warning and then calls the deprecated function.

**Comment by clauswilke** [comment]
@yutannihilation I know. But @Takashiidobe talked about modifying the auto-generated wrapper functions, that's why I brought it up. I also wanted to highlight that even if a package doesn't normally provide wrappers of wrappers, these can be added after the fact for the deprecated functions, as I outlined.

**Comment by yutannihilation** [comment]
I see, thanks for clarifying.

**Comment by Ilia-Kosenkov** [comment]
However, there are packages that expose wrappers to the consumers, e.g.
https://github.com/r-lib/rlang/blob/f1bf5187c47182c723944a361e1c5c8b3f3a3000/R/quo.R#L71
This is the equivalent of decorating rust function in extendr with `/// @export`.
So there is a use case for thin wrappers around native code, especially if these functions can be used in hot paths (frequently called in R code to do small bits of work).

**Comment by yutannihilation** [comment]
rlang doesn't use such framework like cpp11. These are the hand-crafted wrapers, so it's not the case. But I agree with you in general.

**Comment by Ilia-Kosenkov** [comment]
> rlang doesn't use such framework like cpp11
Oh, I had no idea, you are correct. Yeah, I'd recommend to wrap unsafe/ffi code in some user-facing functions rather than directly exposing it to users. Examples used by `{rextendr}` are useful because they allow you to immediately test whatever code you wrote in rust, but in a mature package, it's better to provide an additional layer of wrappers.
